### PR TITLE
Add TraceState to SpanContext in API

### LINKED ISSUE #1302: API: Add TraceState to SpanContext
The spec mentions TraceState under SpanContext:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#spancontext

The previous issue that was filed for this was closed, mentioning that tracestate is not a part of the spec anymore.
I'm not sure why this was done, as an open PR from the spec was linked - and that PR did not end up getting merged into the actual spec

https://github.com/open-telemetry/opentelemetry-go/issues/74

### PR DESCRIPTION
**Pull request by matej-g**
The specification on `SpanContext`requires that `TraceState` is [part of the of the tracing API](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracestate).

This PR adds `TraceState` to `SpanContext`, implements operations according to the spec and adjusts some related parts of the code base (OTLP span transform, SDK test, `stdout` exporter test).

Some parts based on the previously umerged PR https://github.com/open-telemetry/opentelemetry-go/pull/81.

Does not include changes to propagators, however, if required they can be done as part of this PR as well.

Resolves #1302.

### CODE REVIEW

File: trace/trace.go
Comment (Aneurysm9) [inline trace/trace.go:242]:
```suggestion
		return *ts, errInvalidTraceStateKeyValue
```

Should this return the existing `TraceState` rather than an empty value?

File: trace/trace.go
Comment (Aneurysm9) [inline trace/trace.go:268]:
```suggestion
		return *ts, errInvalidTraceStateKeyValue
```

Should this return the existing TraceState rather than an empty value?

File: trace/trace.go
Comment (jmacd) [inline trace/trace.go:323]: Can this be a struct, not a pointer? It should simplify the code a bit.

Comment (matej-g) [reply trace/trace.go:323]: I think it could be. It would require some extra work, as `SpanContext` would no longer be comparable (as at some test utils we use it as [map key](https://github.com/open-telemetry/opentelemetry-go/blob/master/oteltest/span.go#L50) or in SDK [in comparison](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/trace/span.go#L316)), but seems like its doable just fine.

File: trace/trace.go
Comment (MrAlias) [inline trace/trace.go:316]: Can we add an `Equal` method to the `TraceState` type and similarly an `Equal` method to the `SpanContext` type? Each method would take another value of the same type and return a bool based on their equivalence. It seems like we are doing a lot of comparisons and it would clean things up a bit.

File: trace/trace.go
Comment (MrAlias) [inline trace/trace.go:184]: this is a nitpick, but this addition do this file might be easier to read in its own tracestate-specific file.

File: trace/trace.go
Comment (MrAlias) [inline trace/trace.go:50]: If I understand [the specification](https://www.w3.org/TR/trace-context-1/#key) this can start with a digit as well:

> key = ( lcalpha / DIGIT ) 0*240( lcalpha / DIGIT / "_" / "-"/ "*" / "/" ) "@" lcalpha 0*13( lcalpha / DIGIT / "_" / "-"/ "*" / "/" )

File: trace/trace.go
Comment (MrAlias) [inline trace/trace.go:299]: We should probably make a copy of the KeyValue slice here, otherwise, since a slice is a reference type, modification of the original slice will modify the returned TraceState.

File: trace/trace_test.go
Comment (MrAlias) [inline trace/trace_test.go:659]: I feel like this should take into consideration the [truncation scheme of the tracestate](https://www.w3.org/TR/trace-context-1/#tracestate-limits):

> In a situation where tracestate needs to be truncated due to size limitations, the vendor MUST truncate whole entries. Entries larger than 128 characters long SHOULD be removed first. Then entries SHOULD be removed starting from the end of tracestate.

This would mean that `keyx=valux` would be added to the beginning and the last entry would be dropped based on my understanding. I think returning an error as well as implementing this logic seems appropriate.

File: trace/trace_test.go
Comment (MrAlias) [inline trace/trace_test.go:741]: This is called a lot. Can we call it once and save it as a file scoped var?

File: trace/trace_test.go
Comment (MrAlias) [inline trace/trace_test.go:702]: Similar here with regards to truncation. Should this return a TraceState with `keyx=valX` truncated as well as the included error?

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:316]: Agreed, this would be a useful addition considering that with last change there is more code required for comparison. I'll add them.

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:50]: Good catch :+1:

File: trace/trace_test.go
Comment (matej-g) [reply trace/trace_test.go:659]: My understanding of the specification was this applies only to 'size limitations', considering the context of the section (https://www.w3.org/TR/trace-context-1/#tracestate-limits). i.e. situations where it is necessary to truncate the header due to length and not due to number of members in the list. However, now I'm unsure what would be the expected behavior with regards to exceeding max number of members.

Comment (MrAlias) [reply trace/trace_test.go:659]: That's a good point. I'm not entirely sure what behavior is expected either now that you point this out. I'm going to ask around with people who developed that specification to see if I can get clarity.

Comment (MrAlias) [reply trace/trace_test.go:659]: Opened https://github.com/w3c/trace-context/issues/440

**Comment by MrAlias** [comment]
I'm looking at [this](https://github.com/lightstep/tracecontext.go/blob/1757c391b1acf4147823503f13e003115ea4e5df/tracestate/package.go#L48-L49) reference implementation of the tracestate and I'm wondering if we want to take a similar approach the `TraceState` type. Having it itself be a slice would allow the user of this to interact with the data structure in idiomatic Go ways instead of requiring us to build out all the getters/setters.

I think the only outstanding issue would be around how to handle invalid (e.g. too many members, invalid members) tracestate when serializing back into the header.

**Comment by matej-g** [comment]
> Having it itself be a slice would allow the user of this to interact with the data structure in idiomatic Go ways instead of requiring us to build out all the getters/setters.

Unless I'm missing something, we would not want the user to directly interact with the slice, as [the API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracestate) puts some limitations on `TraceState` manipulation (e.g. is immutable, must be valid at all times, must validate input params etc.), which we can control with get / add / delete methods.

**Comment by MrAlias** [comment]
> Unless I'm missing something, we would not want the user to directly interact with the slice, as [the API spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracestate) puts some limitations on `TraceState` manipulation (e.g. is immutable, must be valid at all times, must validate input params etc.), which we can control with get / add / delete methods.

Ah, makes sense. Thanks for the link to the specification, TIL. :+1:

**Comment by matej-g** [comment]
@MrAlias thanks for the feedback :heart:, I updated parts which are clear, waiting for clarifications on the rest.

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:50]: Looks like this is also not 100% clear, looking at your issue at https://github.com/w3c/trace-context/issues/441. Let's see after clarification.

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:184]: Do you mean by that to put `TraceState` into a separate file?

Comment (MrAlias) [reply trace/trace.go:184]: Yeah, `TraceState` and related things. But, we should probably do this in a separate PR as there are many good splits that can happen.

File: trace/trace.go
Comment (MrAlias) [reply trace/trace.go:50]: True. I'm fine with leaving this as is. Given I opened the w3C issue I'll be sure to know what the final decision is and can open an follow on issue if needed.

File: trace/trace_test.go
Comment (MrAlias) [reply trace/trace_test.go]: Same here regarding resolving this and I'll open an follow on issue/PR when the w3c issue is resolved.

File: trace/trace_test.go
Comment (MrAlias) [inline trace/trace_test.go:105]: Could this use the new `IsEqualWith` method?

**Review by MrAlias — APPROVED** [review]

File: trace/trace_test.go
Comment (matej-g) [reply trace/trace_test.go]: Done :+1:

**Review by Aneurysm9 — APPROVED** [review]

File: trace/trace.go
Comment (jmacd) [inline trace/trace.go:201]: I was comparing with the source in `propagation/baggage.go` (`Baggage.Inject()`) and wonder if the differences are intentional. These in theory are the same format, but one is escaped. Should this use TrimSpace()?

File: trace/trace.go
Comment (jmacd) [inline trace/trace.go:218]: This means that the order of keys is significant. Can you make this a test-only function, instead of a TraceState/SpanContext method? The only call I see from non-test code tests against the empty span context, which has no trace state. I'd like to not introduce a public method named `IsEqualWith()` for users of the SDK to depend with this kind of order dependency.

File: trace/trace.go
Comment (jmacd) [inline trace/trace.go:252]: Feel free to ignore:

A lot of the new code to implement TraceState feels like it exists elsewhere in the code base. For example, I think you could implement trace state with a `label.Set`, which maintains a unique key value set _and provides an order-independent equality mechanism_.

File: trace/trace.go
Comment (jmacd) [reply trace/trace.go]: The use of a pointer in this case means that using a SpanContext with a map key can never be equal to another. Do users want a comparable SpanContext like this? Is this use of SpanContext as a map key for more than testing purposes? 

In my feedback elsewhere in this PR, it's occurred to me that some of the `label/set.go` code could be used to implement trace state. At that point we could offer users a form of span context that works as a map key and supports order-independent trace state equality. I won't say this is possible w/o modifications/extensions in the label.Set code, but it's appealing to me nevertheless (e.g., the `label.Set` method supports iteration but is not directly usable as a map key, `label.Distinct` works as a map key but does not support iteration, and there's no support for deletion as a primitive).

**Review by jmacd — APPROVED** [review]
None of my feedback should be considered blocking. If you agree with the suggestion to use `label.Set` and its map-key equality support, that could be a TODO. The suggestion to move `IsEqualWith()` into a test package feels important to do now. I'll approve and let the maintainers decide. :-)

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:201]: As `TraceState` can include only valid key-value entries according to [spec](https://www.w3.org/TR/trace-context/#list-members) (i.e. this is enforced via regexps), would escaping / trimming be required for representing `TraceState` as a string? 

My understanding is that for example `key1=valX&valY` is a valid member and I would expect the string representation to include 'unescaped' string instead of representing it as `valX%26valY` (as this can theoretically also be a valid entry according to the spec). I would not be able to tell what is the actual value of `key1`, whether the escaped or unescaped version. The same would go for leading / trailing space - unless these are allowed by the spec, they should not find their way into entries of a valid `TraceState`.

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:252]: If you're fine with marking it TODO as you mention in the other comment, I'd leave it at that.

File: trace/trace.go
Comment (matej-g) [reply trace/trace.go:218]: I see your point. Yes, there are only two uses outside tests (in SDK and in `oteltest` package), but it should not be a problem to work around it. I will then move these methods to trace test package, hope everyone is fine with that.

**Review by jmacd — APPROVED** [review]

**Review by XSAM — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Pull request merged**
